### PR TITLE
add maxCoins estimation

### LINKED ISSUE #152: Estimate max coins in the deploySC function

### PR DESCRIPTION
**Pull request by Elli610**
closes #152

### CODE REVIEW

File: packages/sc-deployer/src/index.ts [hunk 1]
```diff
@@ -216,17 +230,32 @@ const pollAsyncEvents = async (
...
   fee = 0n,
   maxGas = 1_000_000n,
   wait = false,
 ): Promise<IDeploymentInfo> {
+  // check if maxCoins is set
+  if (!maxCoins) {
+    try{
+      // estimate the maxCoins value
+      maxCoins = estimateMaxCoin(
+        readFileSync(path.join(__dirname, '..', 'build', '/main.wasm')),
```
Comment (Ben-Rey) [inline packages/sc-deployer/src/index.ts:250]: wont always be main.wasm
it should be data from contracts array

File: packages/sc-deployer/src/index.ts [hunk 2]
```diff
@@ -176,6 +176,20 @@ const pollAsyncEvents = async (
...
+ *
+ * @param contractByteCode - The byte code of the smart contract to be deployed.
+ * @param transactionOperationCost - The transaction operation cost in the smallest massa unit. 
+ * (it should be the same as the 'coins' parameter used in the deploySC function).
+ *
+ * @returns The estimated value of the maxCoins value in the smallest massa unit.
+ */
+function estimateMaxCoin(contractByteCode: Buffer, transactionOperationCost: string) {
+  const byteCodeSize = BigInt(contractByteCode.length);
+  return BigInt(10250000) + byteCodeSize * BigInt(250000) + BigInt(transactionOperationCost);
```
Comment (Ben-Rey) [inline packages/sc-deployer/src/index.ts:190]: BigInt(10250000) and other values should be commented constants variables

File: packages/sc-deployer/src/index.ts [hunk 3]
```diff
@@ -176,6 +176,20 @@ const pollAsyncEvents = async (
...
+/**
+ * Estimates the value of the maxCoins maximum number of coins to that should be used while deploying a smart contract.
+ *
+ * @param contractByteCode - The byte code of the smart contract to be deployed.
+ * @param transactionOperationCost - The transaction operation cost in the smallest massa unit. 
+ * (it should be the same as the 'coins' parameter used in the deploySC function).
+ *
+ * @returns The estimated value of the maxCoins value in the smallest massa unit.
+ */
+function estimateMaxCoin(contractByteCode: Buffer, transactionOperationCost: string) {
```
Comment (Ben-Rey) [inline packages/sc-deployer/src/index.ts:188]: transactionOperationCost should be coins or coinsSent, something like that

File: packages/sc-deployer/src/index.ts [hunk 4]
```diff
@@ -216,17 +230,32 @@ const pollAsyncEvents = async (
...
+  // check if maxCoins is set
+  if (!maxCoins) {
+    try{
+      // estimate the maxCoins value
+      maxCoins = estimateMaxCoin(
+        readFileSync(path.join(__dirname, '..', 'build', '/main.wasm')),
+        contracts.reduce((acc, contract) => acc + contract.coins, 0n).toString(),
+      );
+    } catch (ex) {
+      console.log('Error estimating maxCoins value. Using default value.');
```
Comment (Ben-Rey) [inline packages/sc-deployer/src/index.ts:254]: Why it would fail ?

File: packages/sc-deployer/src/index.ts [hunk 5]
```diff
@@ -216,17 +230,32 @@ const pollAsyncEvents = async (
...
+      maxCoins = 4_200_000_000n;
```
Comment (Ben-Rey) [inline packages/sc-deployer/src/index.ts:255]: Why did you choose this value ?

File: packages/sc-deployer/src/index.ts (hunk 4 above)
Comment (Elli610) [reply packages/sc-deployer/src/index.ts:254]: if the main.wasm isn't accessible

**Review by Ben-Rey — APPROVED** [review]

**Review by Thykof — APPROVED** [review]

**Pull request merged**